README.md
## <img src="gui/files/retro_16x16.png" alt="retro" /> retro - Apple II Emulator 

|                                 |                                 |                                             |                                             |
|---------------------------------|---------------------------------|---------------------------------------------|---------------------------------------------|
| ![Aztec](gui/files/aztec_1.png) | ![Aztec](gui/files/aztec_2.png) | ![Lode Runner](gui/files/lode_runner_1.png) | ![Lode Runner](gui/files/lode_runner_2.png) |

### Compatibility

The Retro Apple II Emulator has been developed and tested with:
* `Ubuntu 22 (x86_64)`
* `X11, X.Org version: 1.21.1.4`
* `OpenGL 2.1` compatible graphics drivers (mesa, nvidia)
* `Go 1.21`, `gcc 11.4`

### What's In The Retro Box?
* Board
  * CPU MOS 6502, variable speed (from own [cultivation](https://github.com/dtgorski/m6502))
  * RAM 48KB (+ 16KB Language Card)
  * Applesoft Basic ROM¹
  * Apple Disk II Interface ROM¹
* Display modes (using a 280 x 192 pixel resolution):
  * Default (Text 40 x 24, monochrome)
  * Low Resolution (LoRes 40 x 48, 16 colors)
  * High Resolution (HiRes 140 x 192, 6/8 colors)
  * Mixed (in all modes, the lower 32 pixel rows show four lines of monochrome Text)
* Cards in slots
  * #0: Language Card (16KB RAM, banked)
  * #6: Apple Disk II interface with two diskette drives (16 sector)
* Implementation specific
  * ```CTRL-SHIFT-R``` triggers a reset
  * ```CTRL-V``` pastes the clipboard content
  * persistent configuration, especially convenient for color calibration
  * the source code - if you are interested - is fairly easy to comprehend

¹ included for educational purposes, not owned or licensed by this project

### What Is Missing?
* Speaker sound, in the first place
* Creating and writing of disk images
* Double HiRes and 80x24 character resolution

### Configuration
A configuration file `retro.config.yml` exists in the project directory and is distributed
along with the `retro` executable binary.

### Download

The releases are listed on the right side. It is recommended to use the latest release, especially when
you want to report issues. 

### Build And Run

A Go compiler is required. The code is a CGO application, that is, a Go program with linked C libraries.
Hence before you `make build` or `make run`, following non-Go packages must be additionally installed on your system:

```sh
$ sudo apt install  \
    libxi-dev       \
    libglx-dev      \
    libgl-dev       \
    libx11-dev      \
    libxrandr-dev   \
    libxcursor-dev  \
    libxinerama-dev \
    libxxf86vm-dev  \
    pkg-config      \
    make            \
    gcc
```  
Then ...
```sh
$ git clone https://github.com/dtgorski/retro.git    
$ cd retro && make run # see @dev below    
```
**The first build (when Go builds up caches) might take a bit longer as usual.**

### Usage

```
$ retro -h

Usage: retro [options]
Retro Apple II Emulator v0.0.0

The default name of the configuration file is "retro.config.yml".
At program start, following configuration locations are checked:
first the directory, where the executable resides; then outside,
in ~/.config/retro/, /usr/local/etc/retro/, and /etc/retro/

When no disk image is provided (with -1), the emulator will boot
into the Applesoft BASIC prompt. Additionally, the ROM(s) of the
virtual Apple Disk II Interface will not be mounted into memory.

Command line options override their configuration counterparts. 

Options:
    -1 <path/to/image>
         The APPLE DISK II .dsk image to insert into Drive 1.
         Path can be a HTTP URL. Fetched images are not saved.
    
    -2 <path/to/image>
         The APPLE DISK II .dsk image to insert into Drive 2.
         Path can be a HTTP URL. Fetched images are not saved.

    -c <path/to/config>
         Path to an alternative configuration file. 

    -m <cpu-clock-in-mhz>
         MHz speed of the CPU clock. Usual clock settings are
         0.98 (1MHz) and 3.58 (4MHz). The first is the default.

    -z <window-zoom [1..n]>
         Window magnification. A zoom factor of 1 is equivalent
         to the Apple II native resolution of 280 x 192 pixels.
         Default value: 3 

More options:
    -h  Display this usage help and exit.
    -v  Print program version and exit.
```

### Apple II Diskette Images
Supported formats for disk images:
* Apple II DSK 16 Sector format (`.dsk`, usually 140KB in size)

You can find these images by searching the Web for: `apple ii dsk download`.
Please be aware that these images may be subject to copyright restrictions.

```
$ wget https://<...>/Name.dsk
$ retro -1 Name.dsk
```
...or...
```
$ retro -1 https://<...>/Name.dsk
```

### @dev
Try ```make```:
```
$ make

 make help           Displays this list
 make clean          Removes build/test artifacts
 make dist-clean     Removes Go caches and artefacts debris
 make build          Builds executable binary
 make run            Runs the application
 make tidy           Formats source files, cleans go.mod
```

### Internal Structure Overview
```
├── context
│   └── Context interface {
│           Deadline() (deadline time.Time, ok bool)
│           Done() <-chan struct {}
│           Err() error
│           Value(key interface {}) interface {}
│       }
├── error interface {
│       Error() string
│   }
├── io
│   └── Reader interface {
│           Read(p []byte) (n int, err error)
│       }
├── retro
│   ├── emu
│   │   ├── device
│   │   │   ├── builtin
│   │   │   │   ├── Keyboard struct {}
│   │   │   │   └── Paddles struct {}
│   │   │   ├── diskette
│   │   │   │   ├── Card struct {}
│   │   │   │   ├── Decoder struct {}
│   │   │   │   ├── Drive struct {}
│   │   │   │   ├── Encoder struct {}
│   │   │   │   ├── Image struct {}
│   │   │   │   ├── Reader interface {
│   │   │   │   │       Read() byte
│   │   │   │   │   }
│   │   │   │   ├── Track struct {}
│   │   │   │   └── TrackReader struct {}
│   │   │   ├── language
│   │   │   │   └── Card struct {}
│   │   │   └── render
│   │   │       ├── Driver struct {}
│   │   │       ├── Font struct {}
│   │   │       ├── HiRes struct {}
│   │   │       ├── LoRes struct {}
│   │   │       ├── Mode byte
│   │   │       ├── Modes map[render.Mode]render.Renderer
│   │   │       ├── Renderer interface {
│   │   │       │       Render(page byte, canvas []byte, flash bool)
│   │   │       │   }
│   │   │       └── Text struct {}
│   │   ├── input
│   │   │   ├── CursorPos struct {}
│   │   │   ├── KeyInput struct {}
│   │   │   ├── KeyMap struct {}
│   │   │   └── MouseButton struct {}
│   │   ├── memory
│   │   │   ├── Bus interface {
│   │   │   │       Read(lo byte, hi byte) byte
│   │   │   │       Write(lo byte, hi byte, b byte)
│   │   │   │   }
│   │   │   ├── DMA interface {
│   │   │   │       DMA() []byte
│   │   │   │   }
│   │   │   ├── Device interface {
│   │   │   │       Read(lo byte, hi byte) (byte, bool)
│   │   │   │       Reset()
│   │   │   │       Slot(num byte)
│   │   │   │       Write(lo byte, hi byte, b byte) bool
│   │   │   │   }
│   │   │   ├── Manager struct {}
│   │   │   └── Memory interface {
│   │   │           memory.Bus
│   │   │           memory.DMA
│   │   │           DMA() []byte
│   │   │           Load(addr uint16, r io.Reader) error
│   │   │           MustLoad(addr uint16, r io.Reader)
│   │   │           Read(lo byte, hi byte) byte
│   │   │           Write(lo byte, hi byte, b byte)
│   │   │       }
│   │   └── virtual
│   │       ├── Bridge struct {}
│   │       ├── CPU interface {
│   │       │       PC(lo byte, hi byte)
│   │       │       PCH() byte
│   │       │       PCL() byte
│   │       │       Reset()
│   │       │       Step() (cycles uint, err error)
│   │       │   }
│   │       ├── Channels struct {}
│   │       └── Machine struct {}
│   └── gui
│       ├── Properties struct {
│       │       Width int
│       │       Height int
│       │       Title string
│       │   }
│       └── Window struct {}
└── time
    └── Time struct {}
```

---

### A Personal Note
> *"As many others of my generation, I have started my computer career in 1983 with an Apple II machine.
> Those early days were filled with the excitement of exploration and discovery, as I marveled at
> the possibilities that this revolutionary piece of technology offered.*
>
> *Now, after four decades, I embark on a nostalgic journey, choosing to revisit and relive the memories
> of my youth with an Apple II emulator, that not only resurrects the classic experience, but also pays
> homage to the era that sparked my passion for computing.*
>
> *I hope this emulator brings back fond memories for you as it has for me."*
>
> -- Daniel T. Gorski, author
 
### License
[MIT](https://opensource.org/licenses/MIT) - © dtg [at] lengo [dot] org


CHANGELOG.md
### v0.1.0
* First milestone.


emu/config/config.go
// MIT License · Daniel T. Gorski · dtg [at] lengo [dot] org · 12/2023

package config

import (
	"fmt"
	"gopkg.in/yaml.v3"
	"os"
	"path/filepath"
	"strings"
)

type (
	// Config is the main configuration structure.
	Config struct {
		Version string
		Window  `yaml:"window"`
		CPU     `yaml:"cpu"`
		Disk    `yaml:"disk"`
		Render  `yaml:"render"`
	}

	// Window ...
	Window struct {
		Title string `yaml:"title"`
		Zoom  int    `yaml:"zoom"`
	}

	// CPU ...
	CPU struct {
		MHz float64 `yaml:"mhz"`
	}

	// Disk ...
	Disk struct {
		Drive1 string `yaml:"drive-1"`
		Drive2 string `yaml:"drive-2"`
	}

	// Render ...
	Render struct {
		Mono  `yaml:"mono"`
		LoRes `yaml:"lores"`
		HiRes `yaml:"hires"`
	}

	// Mono ...
	Mono struct {
		Color int `yaml:"color"`
	}

	// LoRes ...
	LoRes struct {
		Colors []int `yaml:"colors"`
	}

	// HiRes ...
	HiRes struct {
		Colors []int `yaml:"colors"`
	}
)

// DefaultConfig is a working configuration.
var DefaultConfig = &Config{

	Window: Window{
		Title: "RETRO Apple II    │    6502 @ %.2f MHz    │    RESET:  CTRL + SHIFT + R",

		// As the resolution of an Apple II is 280x192px,
		// the native presentation mode may be too small.
		Zoom: 3,
	},
	CPU: CPU{
		// Usual clock settings are 0.98 and 3.58 (aka 1MHz and 4MHz).
		MHz: 0.98,
	},

	// File paths of "inserted" Disk 1 and Disk 2 images.
	// Paths can be HTTP URLs. Fetched images are not saved.
	// Using the -1 and -2 options overrides this setting.
	Disk: Disk{},

	Render: Render{
		Mono: Mono{
			// The color of the monochrome text.
			Color: 0x00B500FF,
		},
		LoRes: LoRes{
			Colors: []int{
				0x000000FF, // 0 Black
				0x9A2110FF, // 1 Magenta
				0x3C22A5FF, // 2 Dark Blue
				0xA13FB7FF, // 3 Purple
				0x07653EFF, // 4 Dark Green
				0x7B7E80FF, // 5 Dark Gray
				0x308FE3FF, // 6 Medium Blue
				0xADD8E6FF, // 7 Light Blue
				0x8B4513FF, // 8 Brown
				0xC77028FF, // 9 Orange
				0x7B7E80FF, // A Gray
				0xF39AC2FF, // B Pink
				0x2FB81FFF, // C Green
				0xFFFF80FF, // D Yellow
				0x6EE1C0FF, // E Aqua
				0xFFFFFFFF, // F White
			},
		},
		HiRes: HiRes{
			Colors: []int{
				0x000000FF, // 0 Black
				0xA13FB7FF, // 2 Purple
				0x2FB81FFF, // 1 Green
				0xFFFFFFFF, // 3 White
				0x000000FF, // 4 Black
				0x308FE3FF, // 6 Med Blue
				0xC77028FF, // 5 Orange
				0xFFFFFFFF, // 7 White
			},
		},
	},
}

// FromYML fills the Config object from a YAML file.
// Overwrites existing values in this configuration.
func (c *Config) FromYML(name string) error {
	path := c.findConfig(name)

	file, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("config not loadable: %w", err)
	}
	err = yaml.Unmarshal(file, c)
	if err != nil {
		return fmt.Errorf("syntax error %s: %w", path, err)
	}

	// Ensure list sizes, when config is incomplete.
	lores := &c.Render.LoRes
	lores.Colors = append(lores.Colors, make([]int, 0x10)...)
	lores.Colors = lores.Colors[:0x10]

	hires := &c.Render.HiRes
	hires.Colors = append(hires.Colors, make([]int, 0x08)...)
	hires.Colors = hires.Colors[:0x08]

	return nil
}

func (*Config) findConfig(name string) string {
	exec, _ := os.Executable()
	self := filepath.Dir(exec)

	locations := []string{
		"%s",
		fmt.Sprintf("%s/%%s", self),
		"~/.config/retro/%s",
		"/usr/local/etc/retro/%s",
		"/etc/retro/%s",
	}

	// For development, when executable resides in /bin.
	if strings.HasSuffix(self, "/bin") {
		locations = append([]string{"../%s"}, locations...)
	}

	for _, location := range locations {
		path := fmt.Sprintf(location, name)

		info, err := os.Stat(path)
		if err != nil || info.IsDir() {
			continue
		}
		return path
	}
	return name
}
